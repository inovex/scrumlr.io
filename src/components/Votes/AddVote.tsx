type AddVoteProps = {
  addVote: () => void;
};

export const AddVote = (props: AddVoteProps) => (
  <button
    className="votes__add"
    onClick={(e) => {
      e.stopPropagation();
      props.addVote();
    }}
  >
    +
  </button>
);
