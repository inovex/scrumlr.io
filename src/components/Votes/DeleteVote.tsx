type DeleteVoteProps = {
  deleteVote: () => void;
  numberOfVotes: number;
};

export const DeleteVote = (props: DeleteVoteProps) => (
  <button
    className="votes__delete"
    onClick={(e) => {
      e.stopPropagation();
      props.deleteVote();
    }}
  >
    {props.numberOfVotes}
  </button>
);
