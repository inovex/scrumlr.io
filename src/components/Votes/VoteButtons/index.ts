import {AddVoteButton as Add} from "./AddVoteButton";
import {RemoveVoteButton as Remove} from "./RemoveVoteButton";

export const VoteButtons = {
  Add,
  Remove,
};
