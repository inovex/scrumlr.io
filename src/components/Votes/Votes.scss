@import "src/constants/style";
@import "src/components/TooltipButton/TooltipButton.scss";

$votes-button-size: 28px;

.votes {
  display: inline-flex;
  height: var(--note-footer-height, $votes-button-size);
  justify-content: flex-end;
  gap: 6px;
}
