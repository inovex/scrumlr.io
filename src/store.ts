import { createStore, compose } from 'redux';
import { reactReduxFirebase } from 'react-redux-firebase';

import rootReducer from './reducers';
import { firebase as fbConfig } from './config';
import * as firebase from 'firebase';

export default function configureStore(initialState: {}, history?: {}) {
  firebase.initializeApp(fbConfig);

  const composeEnhancers =
    (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
  const createStoreWithMiddleware = composeEnhancers(
    reactReduxFirebase(firebase, {
      userProfile: 'users',
      enableLogging: false
    })
  )(createStore);
  const store = createStoreWithMiddleware(rootReducer);

  // tslint:disable
  if ((module as any).hot) {
    // Enable Webpack hot module replacement for reducers
    (module as any).hot.accept('./reducers', () => {
      const nextRootReducer = require('./reducers');
      store.replaceReducer(nextRootReducer);
    });
  }
  // tslint:enable

  return store;
}
