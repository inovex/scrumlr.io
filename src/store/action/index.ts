import {BoardActionFactory, BoardAction, BoardReduxAction} from "./board";
import {NoteActionFactory, NoteAction, NoteReduxAction} from "./note";
import {VoteActionFactory, VoteAction, VoteReduxAction} from "./vote";
import {ParticipantAction, ParticipantActionFactory, ParticipantReduxAction} from "./participants";
import {ColumnActionFactory, ColumnAction, ColumnReduxAction} from "./column";
import {RequestActionFactory, RequestAction, RequestReduxAction} from "./request";
import {VotingActionFactory, VotingAction, VotingReduxAction} from "./votings";
import {AuthActionFactory, AuthAction, AuthReduxAction} from "./auth";
import {ViewAction, ViewActionFactory, ViewReduxAction} from "./view";
import {ReactionAction, ReactionActionFactory, ReactionReduxAction} from "./reaction";
import {BoardReactionAction, BoardReactionActionFactory, BoardReactionReduxAction} from "./boardReaction";
import {SkinToneAction, SkinToneActionFactory, SkinToneReduxAction} from "./skinTone";

/** This object lists all internal Redux Action types. */
export const Action = {
  ...BoardAction,
  ...ColumnAction,
  ...NoteAction,
  ...ReactionAction,
  ...AuthAction,
  ...ParticipantAction,
  ...RequestAction,
  ...VoteAction,
  ...VotingAction,
  ...ViewAction,
  ...BoardReactionAction,
  ...SkinToneAction,
};

/** Factory or creator class of internal Redux actions. */
export const Actions = {
  ...BoardActionFactory,
  ...ColumnActionFactory,
  ...NoteActionFactory,
  ...ReactionActionFactory,
  ...RequestActionFactory,
  ...AuthActionFactory,
  ...ParticipantActionFactory,
  ...RequestActionFactory,
  ...VoteActionFactory,
  ...VotingActionFactory,
  ...ViewActionFactory,
  ...BoardReactionActionFactory,
  ...SkinToneActionFactory,
};

/** The types of all application internal redux actions. */
export type ReduxAction = {context: {board?: string; user?: string; voting?: string; serverTimeOffset: number}} & (
  | BoardReduxAction
  | ColumnReduxAction
  | NoteReduxAction
  | ReactionReduxAction
  | AuthReduxAction
  | ParticipantReduxAction
  | RequestReduxAction
  | VoteReduxAction
  | VotingReduxAction
  | ViewReduxAction
  | BoardReactionReduxAction
  | SkinToneReduxAction
);
