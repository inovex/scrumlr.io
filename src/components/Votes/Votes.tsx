import {FC} from "react";
import classNames from "classnames";
import _ from "underscore";
import {useAppSelector} from "store";
import {VoteButtons} from "./VoteButtons";
import "./Votes.scss";

type VotesProps = {
  className?: string;
  noteId: string;
  // Aggregate the votes of the child notes
  aggregateVotes?: boolean;
};

export const Votes: FC<VotesProps> = (props) => {
  const voting = useAppSelector((state) => state.votings.open);
  const ongoingVotes = useAppSelector(
    (state) => ({
      note: state.votes.filter((v) => v.voting === state.votings.open?.id || v.voting === state.board.data?.showVoting).filter((v) => v.note === props.noteId).length,
      total: state.votes.filter((v) => v.voting === state.votings.open?.id || v.voting === state.board.data?.showVoting).length,
    }),
    _.isEqual
  );
  const allPastVotes = useAppSelector(
    (state) =>
      (state.votings.past[0]?.votes?.votesPerNote[props.noteId]?.total ?? 0) +
      (props.aggregateVotes
        ? state.notes.filter((n) => n.position.stack === props.noteId).reduce((sum, curr) => sum + (state.votings.past[0]?.votes?.votesPerNote[curr.id]?.total ?? 0), 0)
        : 0)
  );

  /**
   * If there's no active voting going on and there are no casted votes for
   * this note from previous votings, we don't need to render anything.
   */
  if (!voting && allPastVotes === 0) {
    return null;
  }

  return voting || allPastVotes > 0 ? (
    <div role="none" className={classNames("votes", props.className)} onClick={(e) => e.stopPropagation()}>
      {!voting && allPastVotes > 0 && (
        <VoteButtons.Remove noteId={props.noteId} disabled>
          {allPastVotes}
        </VoteButtons.Remove>
      )}
      {voting && ongoingVotes.note > 0 && <VoteButtons.Remove noteId={props.noteId}>{ongoingVotes.note}</VoteButtons.Remove>}
      {voting && <VoteButtons.Add noteId={props.noteId} disabled={ongoingVotes.total === voting.voteLimit || (ongoingVotes.note > 0 && !voting.allowMultipleVotes)} />}
    </div>
  ) : null;
};
