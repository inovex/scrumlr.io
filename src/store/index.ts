import {createStore, applyMiddleware, combineReducers, Dispatch, MiddlewareAPI, AnyAction} from "redux";
import thunk from "redux-thunk";
import {composeWithDevTools} from "redux-devtools-extension";
import {ApplicationState} from "../types/store";
import {ReduxAction} from "./action";
import {boardReducer} from "./reducer/board";
import {noteReducer} from "./reducer/note";
import {usersReducer} from "./reducer/users";
import {passNoteMiddleware} from "./middleware/note";
import {passBoardMiddleware} from "./middleware/board";
import {passBoardJoinConfirmationMiddleware} from "./middleware/boardJoinConfirmation";
import {passColumnMiddleware} from "./middleware/column";
import {passJoinRequestMiddleware} from "./middleware/joinRequest";
import {joinRequestReducer} from "./reducer/joinRequest";

const parseMiddleware = (stateAPI: MiddlewareAPI<Dispatch<AnyAction>, ApplicationState>) => (dispatch: Dispatch) => (action: ReduxAction) => {
  try {
    return dispatch(action);
  } finally {
    passBoardJoinConfirmationMiddleware(stateAPI, dispatch, action);
    passBoardMiddleware(stateAPI, dispatch, action);
    passColumnMiddleware(stateAPI, dispatch, action);
    passNoteMiddleware(stateAPI, dispatch, action);
    passJoinRequestMiddleware(stateAPI, dispatch, action);
  }
};

const rootReducer = combineReducers<ApplicationState>({
  board: boardReducer,
  notes: noteReducer,
  users: usersReducer,
  joinRequests: joinRequestReducer,
});

const store = createStore(rootReducer, composeWithDevTools(applyMiddleware(thunk), applyMiddleware(parseMiddleware)));

export default store;
