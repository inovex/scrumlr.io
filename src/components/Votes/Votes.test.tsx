import {fireEvent, render} from "@testing-library/react";
import store from "store";
import {ActionFactory} from "store/action";
import {Votes} from "./Votes";

const createVotes = (withVotes: boolean, activeVoting: boolean, className?: string) => (
  <Votes
    noteId="test-id"
    className={className}
    votes={
      withVotes
        ? [
            {
              id: "test-id",
              board: "test-board",
              note: "test-note",
              user: "test-user",
              votingIteration: 1,
            },
          ]
        : []
    }
    activeVoting={activeVoting}
  />
);

describe("Votes", () => {
  describe("should render correctly", () => {
    test("with no votes and disabled voting", () => {
      const votes = render(createVotes(false, false));
      expect(votes.container).toMatchSnapshot();
    });
    test("with no votes and active voting", () => {
      const votes = render(createVotes(false, true));
      expect(votes.container).toMatchSnapshot();
    });
    test("with votes and disabled voting", () => {
      const votes = render(createVotes(true, false));
      expect(votes.container).toMatchSnapshot();
    });
    test("with votes and active voting", () => {
      const votes = render(createVotes(true, true));
      expect(votes.container).toMatchSnapshot();
    });
    test("with additional classname", () => {
      const votes = render(createVotes(false, false, "test-classname"));
      expect(votes.container).toMatchSnapshot();
    });
  });

  describe("should dispatch to store on button press", () => {
    const storeDispatchSpy = jest.spyOn(store, "dispatch");

    test("addVote", () => {
      const votes = render(createVotes(true, true));
      fireEvent.click(votes.container.getElementsByClassName("dot-button")[1]);
      expect(storeDispatchSpy).toHaveBeenCalledWith(ActionFactory.addVote("test-id"));
    });

    test("deleteVote", () => {
      const votes = render(createVotes(true, true));
      fireEvent.click(votes.container.getElementsByClassName("dot-button")[0]);
      expect(storeDispatchSpy).toHaveBeenCalledWith(ActionFactory.deleteVote("test-id"));
    });
  });
});
