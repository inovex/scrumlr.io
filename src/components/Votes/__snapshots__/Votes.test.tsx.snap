// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Votes should render correctly with additional classname 1`] = `
<div>
  <div
    class="votes test-classname"
  />
</div>
`;

exports[`Votes should render correctly with no votes and active voting 1`] = `
<div>
  <div
    class="votes"
  >
    <button
      class="dot-button dot-button__add"
    >
      <svg
        class="dot-button__add-icon"
      >
        icon-add.svg
      </svg>
    </button>
  </div>
</div>
`;

exports[`Votes should render correctly with no votes and disabled voting 1`] = `
<div>
  <div
    class="votes"
  />
</div>
`;

exports[`Votes should render correctly with votes and active voting 1`] = `
<div>
  <div
    class="votes"
  >
    <button
      class="dot-button dot-button__delete"
    >
      <span
        class="dot-button__folded-corner"
      />
      <span>
        1
      </span>
    </button>
    <button
      class="dot-button dot-button__add"
    >
      <svg
        class="dot-button__add-icon"
      >
        icon-add.svg
      </svg>
    </button>
  </div>
</div>
`;

exports[`Votes should render correctly with votes and disabled voting 1`] = `
<div>
  <div
    class="votes"
  >
    <button
      class="dot-button dot-button__delete"
      disabled=""
    >
      <span
        class="dot-button__folded-corner"
      />
      <span>
        1
      </span>
    </button>
  </div>
</div>
`;
